core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/XMLRecordReader.java

```diff
@@ -0,0 +1,305 @@
+            break;
+          case XMLStreamConstants.CHARACTERS:
+            if (nodeFound) {
+              xmlRecord.append(reader.getText());
+            }
+            break;
+          case XMLStreamConstants.END_ELEMENT:
+            String nodeNameEnd = reader.getLocalName();
+            if (nodeFound) {
+              //add closing tag
+              xmlRecord.append(OPENING_END_TAG_DELIMITER).append(nodeNameEnd).append(CLOSING_END_TAG_DELIMITER);
+              if (nodeNameEnd.equals(lastNode)) {
+                nodeFound = false;
+                //set flag for XML record is ready to emit.
+                xmlRecordReady = true;
+              }
+            }
+            nodeLevel--;
+            break;
+          case XMLStreamConstants.START_DOCUMENT:
+            break;
+        }
+        if (xmlRecordReady) {
+          currentValue.put(fileName, xmlRecord.toString());
+          return true;
+        }
+      }
+    } catch (XMLStreamException exception) {
+      throw new IllegalArgumentException(exception);
+    }
+    actionsAfterEOF();
+    return false;
+  }
+
+  /**
+   * Method to append start tag information along with attributes if any
+   */
+  private void appendStartTagInformation(String nodeNameStart, StringBuilder xmlRecord) {
+    xmlRecord.append(OPENING_START_TAG_DELIMITER).append(nodeNameStart);
+    int count = reader.getAttributeCount();
+    for (int i = 0; i < count; i++) {
+      xmlRecord.append(" ").append(reader.getAttributeLocalName(i)).append("=\"").append(reader.getAttributeValue(i))
+        .append("\"");
+    }
+    xmlRecord.append(CLOSING_START_TAG_DELIMITER);
+  }
+
+  /**
+   * Method to take actions after EOF reached.
+   */
+  private void actionsAfterEOF() throws IOException {
+    updateFileTrackingInfo();
+    processFileAction();
+  }
+
+  /**
+   * Method to process file with actions (Delete, Move, Archive ) specified.
+   * @throws IOException - IO Exception occurred while deleting, moving or archiving file.
+   */
+  private void processFileAction() throws IOException {
+    fileAction = fileAction.toLowerCase();
+    switch (fileAction) {
+      case "delete":
+        fs.delete(file, true);
+        break;
+      case "move":
+        Path tagetFileMovePath = new Path(targetFolder + file.getName());
+        fs.rename(file, tagetFileMovePath);
+        break;
+      case "archive":
+        Path targetFile = new Path(targetFolder + file.getName() + ".zip");
+        FileUtililty.archiveFile(fs, file, targetFile, true);
+        break;
+      default:
+        LOG.warn("No action required on the file.");
+        break;
+    }
+  }
+
+  /**
+   * Method to update temporary file with latest XML processed information.
+   * @throws IOException - IO Exception occurred while writing data to temp file.
+   */
+  private void updateFileTrackingInfo() throws IOException {
+    try (FSDataOutputStream outputStream = fs.create(new Path(tempFilePath + "/" + file.getName() + ".txt"))) {
```

Comment:
Try to avoid creating paths using string concatenation like this.  ``new Path(tempFilePath, file.getName() + ".txt")``